README.md
# magicsack

**magicSack** is a utility for storing secret things either *locally* (on the
user's laptop or workstation) or
*over a distributed set of cooperating servers* or on one of the
*cloud services* or on some combination of the three.

MagicSack is in development using Python 3 on Linux.  It should
work on Windows as well, but no effort will be made to confirm this
until the development version is stable on Linux.

## Technical Details

MagicSack is protected by a user-selected passphrase.  This is hashed
using SHA256 and then the first 256 bits of the hash are used as an AES
key.  When documents are added to the system, the user supplies a unique
name and then the document is AES-encrypted using that key before storage.
It should be possible to generate a different key for each document.  The
index from the user-assigned key to the encrypted document is stored
locally.  The encrypted documents may be stored locally or on one or more
cooperating servers or on a cloud service or on any combination of the
three.  The index is always digitally signed and AES-encrypted before
being stored.

The system generates two 2048-bit RSA keypairs.  One of these, **skPriv**,
is used
for creating digital signatures; specifically it is used for signing
the index.  The other, **ckPriv**, is used for encrypting data while
setting up communications with **peers**, hosts cooperating with
`magicSack` in the storage of documents.

## Command Line

	magicSack [-m DIR_NAME] init [-f] [-u U_Dir]
	magicSack [-m DIR_NAME] destroy
	magicSack [-m DIR_NAME] add      FILE [FILE]*
	magicSack [-m DIR_NAME] list     [FILE [FILE]*]
	magicSack [-m DIR_NAME] show     FILE [FILE]*
	magicSack [-m DIR_NAME] drop     FILE [FILE]*
	magicSack [-m DIR_NAME] addPeer  FQDN[:PORT] [FQDN[:PORT]]*
	magicSack [-m DIR_NAME] listPeer FQDN
	magicSack [-m DIR_NAME] dropPeer FQDN[:PORT] [FQDN[:PORT]]*

Each of these commands asks for the passphrase.  Except in the case
of `init`, if the passphrase is wrong, the command will have no effect.

### init

This command creates a new local instance, storing information under
`.magicsack` in the user's home directory.  If the `.magicsack` directory
aleady exists, the command will fail, unless the command line includes
the `-f` or `--force` option.  Unless otherwise specified documents will be
stored in
`.magicsack/uDir`.  If the `-u` option is present, documents will be stored
in `uDir` instead.

The `init` command asks for a passphrase which then becomes the key
for all information stored.

This command is destructive.  Any existing information under `.magicsack`
will be irrevocably deleted.

If data is to be stored on the cloud, `uDir` **must** be specified and
should be the path to a directory backed up on the cloud.  If that
directory does not exist, `magicSack` will attempt to create it

### destroy

This command erases anything under `.magicsack`.  If `uDir` is not a
subdirectory of `.magicsack`, `uDir` will be erased if the system permits.

This operation is irrevocable.  If the correct passphrase is supplied,
all information on the user's machine relating to the `magicSack`
installation will be erased, unless somehow otherwise protected by
the system.

### add

The `add` command adds the files listed.  These are specified using **globs**,
file name patterns.  So if for example `FILE` is `abc*`, then all local files
whose names begin with `abc` will be added to magicsack.

This is not a destructive operation.  That is, if a file is added, it will be
copied into .magicsack, be the original will be unaffected.

### list

`list` is followed by zero or more globs (file name patterns).  If there
are no globs, then
the entire contents of the magicsack store will be listed.  Otherwise
any file matching the pattern will be listed

The listing contains only information about the document(s) requested,
most importantly the document name (which may be a path) and size,
where the size is the size before encryption.

### show

`show` must be followed by one or more globs (file name patterns).
The contens of any file
whose name or path matches the pattern(s) will be displayed.  In other
words, this information will be sent to `stdout`, which may of course
be directed to a file.

### drop

The `drop` command deletes any items matching the file name pattern(s)
following from the store.  This action is irrevocable.

### addPeer

This command adds a remote host (another machine that can be reached
over the Internet) to the list of cooperating servers.  FQDN may be
either a domain name or an IP address, a **dotted quad** like `1.2.3.4`.
Either may optionally be followed by a port number.

### listPeer

Lists domain name or IP address and port number where the name
matches `FQDN` (a glob).  If `FQDN` is not present on the command line,
all such peers are listed.

### dropPeer

Drops peers matching FQDN from the local list of cooperating peers.  This
should be an eventually destructive act: the remote peers will at some
point remove the data stored there, but there is no guarantee when this
will occur.  (Remember that all data stored remotely is encrypted.)

## Project Status

Pre-alpha.  Some working code, a few tests that succeed.

## On-line Documentation

More information on the **magicsack** project can be found
[here](https://jddixon.github.io/magicsack)


CHANGES
~/dev/py/magicsack/CHANGES

v0.4.13
    2018-03-07
        * update setup.py, tox config files                         SLOC 1478
v0.4.12
    2018-02-21
        * replaced hashlib with xlcrypto.hash                       
        * replaced PBKDF2 with xlcrypto.keyderiv.pbkdf2             SLOC 
v0.4.11
    2018-02-19
        * xlattice.crypto becomes xlcrypto                          SLOC 1476
v0.4.10
    2018-01-16
        * AES_BLOCK_SIZE -> _BYTES                                  SLOC 1479
v0.4.9
    2017-12-16
        * pylinting and minor bug fixes                             SLOC 1477
v0.4.8
    2017-10-03
        * add projConfig.toml                                       SLOC 1485
v0.4.7
    2017-09-05
        * pep8; add tox.ini, *requirements.txt                      SLOC 1447
v0.4.6
    2017-07-17
        * rewrite setup.py, MANIFEST.in; edit cfg files             SLOC 1043
v0.4.5
    2017-06-12
        * restructure for pytest                                    SLOC 1393
v0.4.4
    2017-04-24
        * drop dist/* from repo; plus minor edits                   SLOC 1423
v0.4.3
    2017-01-13
        * drop duplicate file testPuzzle.py
        * modify to conform to xlattice_py API changes              SLOC 1206
v0.4.2
    2016-11-29
        * replace listdir with scandir; more pylinting              SLOC 1444
v0.4.1
    2016-10-25
        * changes in xlattice/nlhtree/buildlist APIs                SLOC 1445
v0.4.0
    2016-10-05
        * add pylint, coverage                                      SLOC 1437
v0.3.1
    2016-09-12
        * fix FIX_MEs, limit to SHA2(56)                            SLOC 1276
v0.3.0
    2016-09-02
        * usingSHA1 -> usingSHA : PARTIAL                           SLOC 1288
v0.2.25
    2016-08-25
        * fix setup.py, license                                     SLOC 1278
v0.2.24
    2016-08-18
        * destroy works                                             SLOC 1252
v0.2.23
    2016-08-17
        * init works; general cleanup                               SLOC 1153
v0.2.22
    2016-06-23
        * use xlattice.u.uDir; add ghpDoc/{css,img}/                SLOC 1047
v0.2.21
    2016-05-21
        * add bkp2U, update .dvcz                                   SLOC  860
v0.2.20
    2016-05-10
        * correct use of xlattice.u                                 SLOC  566
v0.2.19
    2016-04-27
        * add .dvcz/projCfg{,.local}                                SLOC  557
v0.2.18
    2016-04-12
        * add .fileMap{,.local}; fix countem                        SLOC  540
v0.2.17
    2016-03-21
        * add and execute auto8it                                   SLOC  548
v0.2.16
    2016-02-24
        * update LICENSE                                            SLOC  526
v0.2.15
    2016-01-28
        * minor edits                                               SLOC  532
v0.2.14
    2015-12-28
        * minor fixes to website
v0.2.13
    2015-11-13
        * minor edits to *.md                                       SLOC  534
v0.2.12
    2015-09-06
        * minor edits to documentation                              SLOC  534
v0.2.11
    2015-06-14
        * changed checkPuzzle signature                             SLOC  508
        * working init subcommand                                   SLOC  534
v0.2.10
    2015-05-31
        * init writes puzzle, BuildList                             SLOC  491
v0.2.9
    2015-05-29
        * add handleAdd                                             SLOC  455
0.2.8
    2015-05-28
        * add generateKey, devisePuzzle, checkPuzzle, and test      SLOC  384
0.2.7
    2015-05-27
        * added top level argument parser                           SLOC  306
0.2.6
    2015-05-19
        * G: index.html updated                                     SLOC  206
        * P: private keys created, encrypted; added --title option  SLOC  249
0.2.5
    2015-05-18
        * init creates .magicsack/cfg.json                          SLOC  206
0.2.4
    2015-05-15
        * pkcs7 functionality moved to xlattice_py/crypto           SLOC  200
        * confirmYorN, passwdStrength moved to xlattice/ui          SLOC  189
0.2.3
    2015-05-14
        * draft password strength assessor
        * loops to accept strength, confirm password                SLOC  285
0.2.2
    2015-05-13
        * skeletal init subcommand                                  SLOC  221
0.2.1
    2015-05-08
        * documentation updates                                     SLOC  185
0.2.0
    2014-12-15
        * convert to Python 3                                       SLOC  185
0.1.1
    2014-11-07
        * add PBKDF2 passphrase verification                        SLOC  173
0.1.0
    2014-10-02
        * initial commit                                            SLOC  152 



































test_puzzle.py
#!/usr/bin/env python3
# testPuzzle.py

""" Verify that the puzzle presented to the user is valid. """

import time
import unittest

from xlattice.crypto import AES_BLOCK_SIZE
from rnglib import SimpleRNG
from magicsack import generate_key, devise_puzzle, check_puzzle


class TestPuzzle(unittest.TestCase):
    """ Verify that the puzzle presented to the user is valid. """

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    def test_key_func(self):
        """ Test key generation. """

        pass_phrase = 'foo' + self.rng.next_file_name(8)
        salt = bytes(self.rng.some_bytes(16))
        key = generate_key(pass_phrase, salt, count=20)
        self.assertTrue(key is not None)
        self.assertEqual(len(key), 2 * AES_BLOCK_SIZE)

    def test_puzzle(self):
        """ Verify that the generated puzzle works as expected. """

        pass_phrase = 'foo' + self.rng.next_file_name(8)
        count = 20    # default is 1000, but this is a test
        salt = bytes(self.rng.some_bytes(16))

        # we want this the return an immutable 64-byte value
        puzzle = devise_puzzle(pass_phrase, salt, self.rng, count)

        ok_, key = check_puzzle(puzzle, pass_phrase, salt, count)
        self.assertTrue(ok_)
        self.assertFalse(key is None)


if __name__ == '__main__':
    unittest.main()


tests/test_puzzle.py
#!/usr/bin/env python3
# testPuzzle.py

""" Verify that the puzzle presented to the user is valid. """

import time
import unittest

from xlcrypto import AES_BLOCK_BYTES
from rnglib import SimpleRNG
from magicsack import generate_key, devise_puzzle, check_puzzle


class TestPuzzle(unittest.TestCase):
    """ Verify that the puzzle presented to the user is valid. """

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    def test_key_func(self):
        """ Test key generation. """

        pass_phrase = 'foo' + self.rng.next_file_name(8)
        salt = bytes(self.rng.some_bytes(16))
        key = generate_key(pass_phrase, salt, iterations=20)
        self.assertTrue(key is not None)
        self.assertEqual(len(key), 2 * AES_BLOCK_BYTES)

    def test_puzzle(self):
        """ Verify that the generated puzzle works as expected. """

        pass_phrase = 'foo' + self.rng.next_file_name(8)
        iterations = 20    # default is 1000, but this is a test
        salt = bytes(self.rng.some_bytes(16))

        # we want this the return an immutable 64-byte value
        puzzle = devise_puzzle(pass_phrase, salt, self.rng, iterations)

        ok_, key = check_puzzle(puzzle, pass_phrase, salt, iterations)
        self.assertTrue(ok_)
        self.assertFalse(key is None)


if __name__ == '__main__':
    unittest.main()


testPKCS7.py
#!/usr/bin/python3

# magicsack/testPKCS7.py

import base64, hashlib, os, time, unittest

from rnglib         import SimpleRNG
from magicsack      import *

class TestPKCS7Padding (unittest.TestCase):

    def setUp(self):
        self.rng = SimpleRNG( time.time() )
    def tearDown(self):
        pass

    # utility functions #############################################
    
    # actual unit tests #############################################
    
    def testPadding (self):

        seven = bytearray(7)
        self.rng.nextBytes(seven)

        fifteen = bytearray(15)
        self.rng.nextBytes(fifteen)

        sixteen = bytearray(16)
        self.rng.nextBytes(sixteen)

        seventeen = bytearray(17)
        self.rng.nextBytes(seventeen)

        padding = pkcs7Padding(seven, AES_BLOCK_SIZE)
        self.assertEquals(len(padding), AES_BLOCK_SIZE-7)
        self.assertEquals(padding[0], AES_BLOCK_SIZE-7)

        padding = pkcs7Padding(fifteen, AES_BLOCK_SIZE)
        self.assertEquals(len(padding), AES_BLOCK_SIZE-15)
        self.assertEquals(padding[0], AES_BLOCK_SIZE-15)

        padding = pkcs7Padding(sixteen, AES_BLOCK_SIZE)
        self.assertEquals(len(padding), AES_BLOCK_SIZE)
        self.assertEquals(padding[0], 16)

        padding = pkcs7Padding(seventeen, AES_BLOCK_SIZE)
        expectedLen = 2*AES_BLOCK_SIZE - 17
        self.assertEquals(len(padding), expectedLen)
        self.assertEquals(padding[0], expectedLen)

        paddedSeven = addPKCS7Padding(seven, AES_BLOCK_SIZE)
        unpaddedSeven = stripPKCS7Padding(paddedSeven, AES_BLOCK_SIZE)
        self.assertEquals(seven, unpaddedSeven)

        paddedFifteen = addPKCS7Padding(fifteen, AES_BLOCK_SIZE)
        unpaddedFifteen = stripPKCS7Padding(paddedFifteen, AES_BLOCK_SIZE)
        self.assertEquals(fifteen, unpaddedFifteen)

        paddedSixteen = addPKCS7Padding(sixteen, AES_BLOCK_SIZE)
        unpaddedSixteen = stripPKCS7Padding(paddedSixteen, AES_BLOCK_SIZE)
        self.assertEquals(sixteen, unpaddedSixteen)

        paddedSeventeen = addPKCS7Padding(seventeen, AES_BLOCK_SIZE)
        unpaddedSeventeen = stripPKCS7Padding(paddedSeventeen, AES_BLOCK_SIZE)
        self.assertEquals(seventeen, unpaddedSeventeen)


if __name__ == '__main__':
    unittest.main()



testPuzzle.py
#!/usr/bin/env python3

# testPuzzle.py

import os
import time
import unittest

from xlattice.crypto import AES_BLOCK_SIZE
from rnglib import SimpleRNG
from magicsack import *


class TestPuzzle (unittest.TestCase):

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    # utility functions #############################################

    # actual unit tests #############################################

    def testKeyFunc(self):
        passPhrase = 'foo' + self.rng.nextFileName(8)
        salt = bytes(self.rng.someBytes(16))
        key = generateKey(passPhrase, salt, count=20)
        self.assertTrue(key is not None)
        self.assertEqual(len(key), 2 * AES_BLOCK_SIZE)

    def testPuzzle(self):
        passPhrase = 'foo' + self.rng.nextFileName(8)
        count = 20    # default is 1000, but this is a test
        salt = bytes(self.rng.someBytes(16))

        # we want this the return an immutable 64-byte value
        puzzle = devisePuzzle(passPhrase, salt, self.rng, count)

        ok, key = checkPuzzle(puzzle, passPhrase, salt, count)
        self.assertTrue(ok)
        self.assertEqual(key is None, False)

if __name__ == '__main__':
    unittest.main()


src/magicsack/__init__.py
# magicsack/__in(it__.py

""" Library for the Magic Sack. """

import binascii
import hashlib
import os

from pbkdf2 import PBKDF2       # note name of package is u/c
from Crypto.Cipher import AES
# from Crypto.Hash import SHA
# from Crypto.PublicKey import RSA

from buildlist import BuildList
from nlhtree import NLHLeaf
from xlcrypto import AES_BLOCK_BYTES
from xlcrypto.padding import add_pkcs7_padding, strip_pkcs7_padding

__all__ = ['__version__', '__version_date__',
           'check_puzzle',
           'devise_puzzle',
           'generate_key',
           'make_named_value_leaf', 'name_from_title',
           'write_build_list', ]

__version__ = '0.4.11'
__version_date__ = '2018-02-19'


class Config(object):
    """ Configuration information. """

    def __init__(self, salt, u_dir):
        self._salt = salt
        self._u_dir = u_dir

    @property
    def salt(self):
        """ Return the salt (some random bytes to confuse hackers). """
        return self._salt

    @property
    def u_dir(self):
        """ Path to the store. """
        return self._u_dir


class MagicSackError(RuntimeError):
    """ Wrapper for errors associated with this package. """
    pass


def name_from_title(title):
    """ convert a title into an acceptable directory name """
    txt = title.strip()           # strip off lealding & trailing blanks
    chars = list(txt)             # atomize the title
    for ndx, char in enumerate(chars):
        if char == ' ':
            chars[ndx] = '_'
        elif char == '(':
            chars[ndx] = '%28'
        elif char == ')':
            chars[ndx] = '%29'
        elif char == '/':
            chars[ndx] = '%2F'
        elif char == '\\':
            chars[ndx] = '%5C'

    return ''.join(chars)


def generate_key(pass_phrase, salt, iterations=1000):
    """
    pass_phrase is a string which may not be empty.  salt is a
    byte array, conventionally either 8 or 16 bytes.  The
    key returned is a 256-bit value.
    """
    if not pass_phrase or pass_phrase == '':
        raise RuntimeError("empty pass_phrase")
    if not salt:
        raise RuntimeError("you must supply a salt")
    # it is also possible to set the hash function used; it defaults
    # to HMAC-SHA1
    return PBKDF2(pass_phrase, salt, iterations=iterations).read(32)


def devise_puzzle(pass_phrase, salt, rng, iterations=1000):
    """
    Create the puzzle that the user has to solve (provide a key for)
    in order to access the Magic Sack.
    """

    key = generate_key(pass_phrase, salt, iterations)
    junk = rng.some_bytes(2 * AES_BLOCK_BYTES)
    iv_ = bytes(junk[:AES_BLOCK_BYTES])
    junk0 = junk[AES_BLOCK_BYTES: AES_BLOCK_BYTES + 8]
    junk2 = junk[AES_BLOCK_BYTES + 8:]

    data = junk0 + salt + junk2
    padded = bytes(add_pkcs7_padding(data, AES_BLOCK_BYTES))

    # DEBUG
    # print("devise_puzzle:")
    # print("  key       %s" % binascii.b2a_hex(key))
    # print("  iv_        %s" % binascii.b2a_hex(iv_))
    # print("  salt      %s" % binascii.b2a_hex(salt))
    # print("  padded    %s" % binascii.b2a_hex(padded))
    # END

    cipher = AES.new(key, AES.MODE_CBC, iv_)
    puzzle = bytes(iv_ + cipher.encrypt(padded))

    return puzzle


def check_puzzle(puzzle, pass_phrase, salt, iterations=10000):
    """
    Determine the key then decipher the puzzle, verifying that
    the copy of the salt embedded in the puzzle is the same as
    the salt from the config file.  Return whether verification
    succeeded.
    """
    key = generate_key(pass_phrase, salt, iterations)

    iv_ = puzzle[:AES_BLOCK_BYTES]
    cipher = AES.new(key, AES.MODE_CBC, iv_)
    decrypted = cipher.decrypt(puzzle[AES_BLOCK_BYTES:])
    # DEBUG
    # print("check_puzzle:")
    # print("  key       %s" % binascii.b2a_hex(key))
    # print("  iv_       %s" % binascii.b2a_hex(iv_))
    # print("  salt      %s" % binascii.b2a_hex(salt))
    # print("  decrypted %s" % binascii.b2a_hex(decrypted))
    # END
    data = strip_pkcs7_padding(decrypted, AES_BLOCK_BYTES)
    soln = bytes(data[8:8 + AES_BLOCK_BYTES])

    return soln == salt, key


# ACTIONS -----------------------------------------------------------
def insert_named_value(global_ns, name, data):
    """
    Pad and encrypt the data, writing the encrypted value into u_dir.
    If successful, return an NLHLeaf.
    """
    u_dir = global_ns.u_dir
    u_path = global_ns.u_path
    hashtype = global_ns.hashtype

    padded = add_pkcs7_padding(data, AES_BLOCK_BYTES)
    iv_ = bytes(global_ns.rng.some_bytes(AES_BLOCK_BYTES))
    cipher = AES.new(global_ns.key, AES.MODE_CBC, iv_)
    encrypted = cipher.encrypt(padded)

    # hash and encrypt the data ---------------------------
    sha = hashlib.sha256()
    sha.update(encrypted)
    bin_hash = sha.digest()
    hex_hash = binascii.b2a_hex(bin_hash).decode('utf-8')

    # DEBUG
    print("len(encrypted) = %d" % len(encrypted))
    print("len(hex_hash)   = %d" % len(hex_hash))
    print("u_path          = %s" % u_path)
    # END

    # add the encrypted data to u_dir -----------------------
    length, hash2 = u_dir.put_data(encrypted, hex_hash)
    if hex_hash != hash2:
        raise MagicSackError(
            "INTERNAL ERROR: content key was '%s' but u returned '%s'" % (
                hex_hash, hash2))
    if len(encrypted) != length:
        raise MagicSackError("length encrypted %d but %d bytes written" % (
            len(encrypted), length))

    return NLHLeaf(name, bin_hash, hashtype)


def make_named_value_leaf(global_ns, name, data):
    """ Given its name and data, insert (name, hash) into the NLHTree. """

    return insert_named_value(global_ns, name, data)


def add_a_file(global_ns, path_to_file, list_path=None):
    """

    Add the contents of a single file to the nlhTree and the content-keyed
    store.  The file is located at 'path_to_file'.  Its name in the NLHTree
    will be 'list_path'.  If list_path is not set, it defaults to path_to_file.

    Return a possibly empty status string.
    """
    key = global_ns.key
    rng = global_ns.rng
    tree = global_ns.tree
    u_dir = global_ns.u_dir
    hashtype = global_ns.hashtype
    status = ''

    # XXX AES KEY IS NOT KNOWN XXX

    if not os.path.exists(path_to_file):
        status = 'file not found: %s' % path_to_file

    if not status:

        # -----------------------------------------------------------
        # NOTE CRITICALLY THIS ASSUMES that the file can be read into memory
        # as a single operation; chunking is not required.
        # -----------------------------------------------------------

        # read, pad, and encrypt the file -----------------

        with open(path_to_file, 'rb') as file:
            data = file.read()
        padded = add_pkcs7_padding(data, AES_BLOCK_BYTES)
        iv_ = rng.some_bytes(AES_BLOCK_BYTES)
        cipher = AES.new(key, AES.MODE_CBC, iv_)
        encrypted = cipher.encrypt(padded)

        # hash the file and add it to u_dir ----------------
        sha = hashlib.sha256()
        sha.update(encrypted)
        hex_hash = sha.hexdigest()

        length, hash_back = u_dir.put_data(encrypted, hex_hash)
        if hash_back != key:
            status =\
                "INTERNAL ERROR: content key was '%s' but u returned '%s'" % (
                    hex_hash, hash_back)
        if not status and len(encrypted) != length:
            status = "length encrypted %d but %d bytes written" % (
                len(encrypted), length)

    if not status:
        # add the file to the NLHTree ---------------------
        if not list_path:
            list_path = path_to_file
        leaf = NLHLeaf(list_path, hex_hash, hashtype)
        tree.insert(leaf, hashtype)

    return status

# BUILD LIST --------------------------------------------------------


def write_build_list(global_ns):
    """ Serialize the BuildList and write it to disk. """

    key = global_ns.key
    magic_path = global_ns.magic_path
    rng = global_ns.rng
    title = global_ns.title
    tree = global_ns.tree
    build_list = BuildList(title, global_ns.sk_, tree)

    # sign build list, encrypt, write to disk -------------
    build_list.sign(global_ns.sk_priv_)
    text = build_list.__str__()
    # DEBUG
    print("BUILD LIST:\n%s" % text)
    # END
    padded = add_pkcs7_padding(text.encode('utf-8'), AES_BLOCK_BYTES)
    iv_ = bytes(rng.some_bytes(AES_BLOCK_BYTES))
    cipher = AES.new(key, AES.MODE_CBC, iv_)
    encrypted = cipher.encrypt(padded)
    path_to_build_list = os.path.join(magic_path, 'bVal')
    with open(path_to_build_list, 'wb') as file:
        file.write(encrypted)


def read_build_list(global_ns):
    """ Read a serialized BuildList from the disk.  """
    key = global_ns.key
    magic_path = global_ns.magic_path
    # rng = global_ns.rng
    # u_path = global_ns.u_path
    hashtype = global_ns.hashtype

    path_to_build_list = os.path.join(magic_path, 'bVal')
    with open(path_to_build_list, 'rb') as file:
        data = file.read()
    iv_ = data[:AES_BLOCK_BYTES]
    ciphertext = data[AES_BLOCK_BYTES]
    cipher = AES.new(key, AES.MODE_CBC, iv_)
    plaintext = cipher.decrypt(ciphertext)
    text = strip_pkcs7_padding(plaintext, AES_BLOCK_BYTES).decode('utf-8')
    build_list = BuildList.parse(text, hashtype)
    if not build_list.verify():
        raise MagicSackError("could not verify digital signature on BuildList")

    global_ns.timestamp = build_list.timestamp
    global_ns.title = build_list.title
    global_ns.tree = build_list.tree

    # retrieve __ckPriv__ and __sk_priv__ hashes from the BuildList, and
    # use these to extract their binary values from u_dir
    # Retrieve any top-level leaf nodes whose names begin with double
    # underscores ('__').  Regard these as reserved names.  For any
    # such keys, add the key/value combination to global_ns, where the
    # value is a hex_hash.

    # NOTE STUB NOTE
